firmware/robot2015/src-kckr/main.c

```diff
@@ -1,13 +1,237 @@
 #include <avr/io.h>
+#include <avr/interrupt.h>
 #include <util/delay.h>
+// TODO: Make this path less explicit!
+#include "kicker_commands.h"
+#include "pins.h"
 
-int main(void) {
-    DDRA = 0xFF;
+#define NO_COMMAND 0
+#define IGNORE_CS 0
 
+#define TIMING_CONSTANT 125
+#define VOLTAGE_READ_DELAY_MS 100
+
+// State of the ATTINY kicking/chipping
+typedef enum { OFF, ON, ACTING } state_type;
+volatile state_type state_ = ON;
+
+// Used to time kicks and chips
+volatile unsigned millis_left_ = 0;
+
+// Used to keep track of current button state
+volatile int kick_db_held_down_ = 0;
+volatile int chip_db_down_ = 0;
+volatile int charge_db_down_ = 0;
+
+uint8_t cur_command_ = NO_COMMAND;
+
+// always up-to-date voltage so we don't have to get_voltage() inside interupts
+uint8_t last_voltage_ = 0;
+// function that does a voltage reading
+uint8_t get_voltage();
+// executes a command coming from SPI
+void execute_cmd(uint8_t, uint8_t);
+
+void main() {
+    /* Port direction - setting outputs */
+    DDRA |= _BV(KICK_PIN) | _BV(CHIP_PIN) |
+            _BV(CHARGE_PIN);  // MISO is handled by CS interrupt
+    // ensure N_KICK_CS is input
+    DDRA &= ~_BV(N_KICK_CS_PIN);
+    // ensure debug buttons are inputs
+    DDRB &= ~_BV(DB_KICK_PIN) & ~_BV(DB_CHIP_PIN) & ~_BV(DB_CHG_PIN);
+    // Which DDRB = 0 and PORTB = 1, these are configured as pull-up inputs
+    PORTB |= _BV(DB_KICK_PIN) | _BV(DB_CHIP_PIN) | _BV(DB_CHG_PIN);
+    // Chip Select Interrupt
+    if (IGNORE_CS)
```

Comment:
That was only added after trying to test over the programming header on the kickerboard which didn't have a chip select pin. I agree that it's a bit dangerous and we shouldn't really need it again so I'll probably just rip it out.